# The main entry point of the workflow.
# After configuring, running snakemake -n in a clone of this repository should successfully execute a dry-run of the workflow.



rule all:
	input:
		'results/COJO/checks/COJO_performed',
		'results/effect_origin/delivery/checks/effect_origin_performed.txt',
		'results/PGS/delivery/glm/PGS_haplotype.txt',
		'results/bilirubin/delivery/lh/total-bilirubin.txt',
		'results/ABO/delivery/ABO-blood-groups.txt',
		'results/colocalization/delivery/pph_jaundice_fets.txt',
		'results/UGT-missense/delivery/jaundice.txt',
		'results/UGT-missense/delivery/jaundice-transmitted.txt',
		'results/conditional_GWAS/checks/GWAS_performed.txt',
		'results/HESS/checks/h2_estimated.txt',
		'results/plots/jaundice-manhattan-mother-child.png',
		'results/plots/fetal-UGT1A4-missense-EAF.pdf',
		'results/plots/fetal-UGT1A4-GA-interaction.pdf',
		'results/plots/adult-bilirubin-PGS-distribution.pdf',
		'results/plots/ABO-alleles.pdf',
		'results/plots/checks/QQ-plot.txt',
		'results/plots/checks/manhattan-jaundice.txt',
		'results/plots/checks/HESS-plot.txt',
		'results/plots/paternal-UGT-alleles.pdf',
		'results/eQTL_catalogue/delivery/pph-jaundice-fets.txt',
		'results/eQTL_catalogue/delivery/SNP-jaundice-fets.txt'

include: "rules/common.smk"
include: "rules/phenotype.smk"
include: "rules/GWAS.smk"
include: "rules/QC_GWAS.smk"
include: "rules/COJO.smk"
include: "rules/effect_origin.smk"
include: "rules/figures.smk"
include: "rules/HESS.smk"
include: "rules/PGS.smk"
include: "rules/total_bilirubin.smk"
include: "rules/ABO.smk"
include: "rules/colocalization.smk"
include: "rules/merge_all_data.smk"
include: "rules/conditional_GWAS.smk"
include: "rules/UGT1A4-missense.smk"
include: "rules/eQTL_catalogue.smk"
